### ISSUE TITLE
Keras performance with OpenBlas Multi-Core

### ISSUE THREAD

**Issue by abhishekpatnia** [fchollet/keras#5601]
**Problem**
I am unable to force Keras + Theano to use more than 1 thread using OpenBLAS. Script used:

```
import numpy as np
from keras.layers import Input, Dense
from keras.models import Model

# this returns a tensor
inputs = Input(shape=(1024000,))

# a layer instance is callable on a tensor, and returns a tensor
x = Dense(256, activation='relu')(inputs)
x = Dense(256, activation='relu')(x)
predictions = Dense(10000, activation='softmax')(x)

model = Model(input=inputs, output=predictions)
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

for i in range(100000):
    print i
    model.predict(np.random.rand(1, 1024000))
```

Command used to run this script

```
THEANO_FLAGS='device=cpu,openmp=True,blas.ldflags=-lopenblas' OMP_NUM_THREADS=12 KERAS_BACKEND=theano python simple_model.py
```

Any help will be great. Thanks.

**System setup**

- CPU with 12 cores
- BLAS using OpenBLAS
- Theano Bleeding Edge installation

**Due Diligence**
- By using htop confirmed that OpenBLAS is using all cores by using Theano check_blas script. 
- By using htop Confirmed that Numpy is using OpenBLAS and all cores by using following script:

```
import numpy as np

def test_eigenvalue():
    i = 500
    data = np.random.rand(i, i)
    np.linalg.eig(data)

for i in range(100000):
    test_eigenvalue()
```

**Comment by unrealwill** [comment]
Hello,

Try using multiple samples : 

```
data = np.random.rand(100, 1024000)
for i in range(100000):
    print i
    model.predict(data)
```

Then your code and your way to start it use multi-core for me.

**Comment by abhishekpatnia** [comment]
Hi @unrealwill 
Thanks for responding. I tried using multiple samples but still seeing 2 threads. 

![screenshot from 2017-03-05 08-56-27](https://cloud.githubusercontent.com/assets/6693858/23589442/fd8bd45e-0181-11e7-8346-745419b115a4.png)


Can you share some details of your system setup. Maybe I am missing something ?

**Comment by unrealwill** [comment]
Keras is just a wrapper of theano. 
The issue is probably with Theano.
Have you succeeded in running openmp with theano.

By looking at 
https://github.com/fchollet/keras/issues/1245

seems that some people may have had success with adding : 
import theano
theano.config.openmp = True

**Comment by abhishekpatnia** [comment]
**Theano Test Code**
```
import numpy as np
from keras.layers import Input, Dense
from keras.models import Model
import os
import theano
import theano.tensor as T

theano.config
print os.getpid()


def test_eigenvalue():
    i = 500
    data = np.random.rand(i, i)
    np.linalg.eig(data)


def test_theano():
    a = theano.shared(np.ones((100, 256), dtype=theano.config.floatX, order="C"))
    b = theano.shared(np.ones((256, 256), dtype=theano.config.floatX, order="C"))
    f = theano.function([], outputs=[T.dot(a, b)])

    for i in range(200000):
        print i
        f()


def test_keras():
    inputs = Input(shape=(5000,), dtype=theano.config.floatX)

    # a layer instance is callable on a tensor, and returns a tensor
    x = Dense(256, activation='relu')(inputs)
    x = Dense(256, activation='relu')(x)
    predictions = Dense(10000, activation='softmax')(x)

    # this creates a model that includes
    # the Input layer and three Dense layers
    model = Model(input=inputs, output=predictions)
    model.compile(optimizer='rmsprop',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    data = np.random.rand(100, 5000).astype(theano.config.floatX)
    for i in range(100000):
        print i
        model.predict(data)

test_theano()
```

**Run command**
```
THEANO_FLAGS='device=cpu,openmp=True,blas.ldflags=-lopenblas' OMP_NUM_THREADS=12 KERAS_BACKEND=theano python simple_model.py
```

**HTOP Output**
![screenshot from 2017-03-05 09-21-10](https://cloud.githubusercontent.com/assets/6693858/23589632/770a96e6-0185-11e7-984b-2f7f2a786071.png)

Yes I am using openmp=True, its part of THEANO_FLAGS.

**Comment by unrealwill** [comment]
I removed my theanorc file.
 ~/.keras/keras.json is {"epsilon": 1e-07, "floatx": "float32", "backend": "theano","image_dim_ordering" : "th"}
test_keras() uses all threads for me.
I'm on ubuntu 16.04, I only have machines with GPU.
I installed keras and theano with pip .
I installed openblas via apt-get libopenblas-dev

**Comment by abhishekpatnia** [comment]
Thanks for you time @unrealwill 

**My system specifications**

- Ubuntu 16.06
- Machine with GPU
- libopenblas-dev 
-Theano installed using PIP

**_I did not use PIP for Keras._**. Maybe that is causing it to be not setup properly. Let me try it and report back.

**Comment by abhishekpatnia** [comment]
No dice. Same behavior. I think I will try to compile the dot product via Keras wrappers and see what that does. 

Thanks again @unrealwill. If you think of anything else let me know.

**Comment by abhishekpatnia** [comment]
Solved it. Concretely, Theano multi-core behavior is based on Tensor sizes. Some information here http://deeplearning.net/software/theano/tutorial/multi_cores.html. Look at the variable openmp_elemwise_minsize. **_In current case, I was not passing in batch size and as such internally the data was fed to Theano with batch size 32 which Theano decided was not good for multi-core processing._** However as soon as I changed the batch size to 100 it triggered multi-core. 

```
import numpy as np
from keras.layers import Input, Dense
from keras.models import Model
import os
import theano
import theano.tensor as T

theano.config
print os.getpid()


def test_eigenvalue():
    i = 500
    data = np.random.rand(i, i)
    np.linalg.eig(data)


def test_theano():
    a = theano.shared(np.ones((100, 256), dtype=theano.config.floatX, order="C"))
    b = theano.shared(np.ones((256, 256), dtype=theano.config.floatX, order="C"))
    f = theano.function([], outputs=[T.dot(a, b)])

    for i in range(200000):
        print i
        f()


def test_keras():
    inputs = Input(shape=(5000,), dtype=theano.config.floatX)

    # a layer instance is callable on a tensor, and returns a tensor
    x = Dense(256, activation='relu')(inputs)
    x = Dense(256, activation='relu')(x)
    predictions = Dense(10000, activation='softmax')(x)

    # this creates a model that includes
    # the Input layer and three Dense layers
    model = Model(input=inputs, output=predictions)
    model.compile(optimizer='rmsprop',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])

    data = np.random.rand(100, 5000).astype(theano.config.floatX)
    for i in range(100000):
        print i
        model.predict(data, batch_size=100)

test_keras()
```

**Run Command**
```
THEANO_FLAGS='device=cpu,openmp=True,blas.ldflags=-lopenblas' OMP_NUM_THREADS=12 KERAS_BACKEND=theano python simple_model.py
```
**HTOP Output**
![screenshot from 2017-03-05 10-46-32](https://cloud.githubusercontent.com/assets/6693858/23590233/1183e69a-0191-11e7-9d6d-743673b35948.png)

**Comment by unrealwill** [comment]
Ok I had only 8 cores on my machine which probably explain the different behaviors.
Glad you solved it :)

**Comment by abhishekpatnia** [comment]
@unrealwill  Thanks for the help.

**Comment by shvthr** [comment]
Hi all,
I have the same issue (I am unable to force Keras + Theano to use more than 1 thread using OpenBLAS.) I have followed the whole procedure here, It works on 

`data = np.random.rand(100, 1024000)
for i in range(100000):
    print i
    model.predict(data)`
but it doesn't work on :
`
THEANO_FLAGS='device=cpu,openmp=True,blas.ldflags=-lopenblas'
OMP_NUM_THREADS=8 
KERAS_BACKEND='theano'
import pandas as pd
import numpy as np
from tqdm import tqdm
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Dropout
from keras.layers.embeddings import Embedding
from keras.layers.recurrent import LSTM, GRU
from keras.layers.normalization import BatchNormalization
from keras.utils import np_utils
import keras.engine.topology 
from keras.layers import TimeDistributed, Lambda
from keras.layers import Convolution1D, GlobalMaxPooling1D
from keras.callbacks import ModelCheckpoint
from keras import backend as K
from keras.layers.advanced_activations import PReLU
from keras.preprocessing import sequence, text

data = pd.read_csv('data/quora_duplicate_questions.tsv', sep='\t')
y = data.is_duplicate.values
tk = text.Tokenizer(nb_words=200000)
max_len = 40
tk.fit_on_texts(list(data.question1.values) + list(data.question2.values.astype(str)))
x1 = tk.texts_to_sequences(data.question1.values)
x1 = sequence.pad_sequences(x1, maxlen=max_len)

x2 = tk.texts_to_sequences(data.question2.values.astype(str))
x2 = sequence.pad_sequences(x2, maxlen=max_len)

word_index = tk.word_index

ytrain_enc = np_utils.to_categorical(y)

embeddings_index = {}
f = open('data/glove.840B.300d.txt')
values_conv = []
for line in tqdm(f):
    values = line.split()
    for item in values:
        try:
            values_conv.append(float(item))
        except: 
            pass
    word = values_conv[0]
    coefs = np.asarray(values_conv[1:], dtype='float32')
    embeddings_index[word] = coefs
f.close()
print('Found %s word vectors.' % len(embeddings_index))
embedding_matrix = np.zeros((len(word_index) + 1, 300))
for word, i in tqdm(word_index.items()):
    embedding_vector = embeddings_index.get(word)
    if embedding_vector is not None:
        embedding_matrix[i] = embedding_vector
`
